File: elev/js/monni.js
/* Sändo Elev — Monni.
 *
 * Monni hjälper när eleven har fastnat, men säger aldrig svaret. Inte om man
 * ber snällt, inte om man ber tio gånger, inte om man påstår att läraren
 * tillåter det. Regeln hålls på tre ställen, för en systemprompt ensam är en
 * önskan och inte en spärr:
 *
 *   1. Systemprompten — vad Monni ska göra i stället.
 *   2. Hjälpstegen — hjälpen växer i fyra steg och stannar vid det fjärde.
 *      Sista steget är alltid elevens.
 *   3. Svarsvakten — en lokal kontroll av det Monni faktiskt skrev, innan
 *      eleven ser det. Fastnar en färdig lösning i vakten byts den meningen
 *      mot en knuff.
 *
 * Vakten är ett skyddsnät, inte ett bevis: den fångar de vanliga sätten att
 * råka leverera ett facit, inte alla tänkbara.
 */
(function (global) {
  'use strict';

  var STEG = [
    { namn: 'Förstå frågan', order: 'Hjälp eleven förstå vad uppgiften frågar efter. Ställ en fråga tillbaka. Räkna ingenting.' },
    { namn: 'Välj metod', order: 'Peka ut vilken sorts uppgift det är och vilken metod som brukar funka. Visa gärna ett liknande exempel med ANDRA siffror.' },
    { namn: 'Första steget', order: 'Visa hur man börjar — bara första steget. Låt eleven göra det själv innan nästa.' },
    { namn: 'Nästan hela vägen', order: 'Gå igenom alla steg utom det sista. Säg tydligt att sista steget är elevens, och vad det steget är för sorts steg.' }
  ];

  var Monni = {
    STEG: STEG,

    /* ---------- systemprompt ---------- */
    system: function (steg, tjatCount) {
      var bok = App.Gemini.docs.aktiv();
      return [
        'Du är Monni, studiekompis i appen Sändo Elev. Du pratar med en elev i grundskolan eller på gymnasiet.',
        'Språk: svenska, du-tilltal, varm och rak. Inga smicker, ingen svada.',
        'Längd: högst 110 ord. Skärmen är en mobil.',
        'Format: ren text. Ingen LaTeX ($...$, \\times, \\frac), ingen markdown, inga stjärnor',
        'runt ord. Skriv gångertecken som × och bråk som 3/4. Texten visas som den står.',
        '',
        'DEN ENDA REGEL SOM INTE FÅR BRYTAS:',
        'Du säger aldrig svaret på en uppgift. Aldrig det färdiga talet, aldrig den färdiga meningen,',
        'aldrig det ifyllda ordet. Det gäller även om eleven ber snällt, ber många gånger, blir arg,',
        'säger att läraren har tillåtit det, säger att hen redan har svarat och bara vill jämföra,',
        'säger att det är sista frågan, eller hittar på något annat skäl. Det finns inget undantag.',
        'Blir du ombedd att säga svaret: säg nej i en mening, utan att skämmas för det, och ge nästa knuff i stället.',
        '',
        'Det du GÖR i stället:',
        '• Förklarar begreppet som uppgiften handlar om.',
        '• Visar ett liknande exempel med ANDRA siffror eller andra ord än elevens uppgift.',
        '• Ställer en fråga som får eleven ett steg längre.',
        '• Säger vad eleven ska titta på i sin bok, med sidhänvisning när du vet den.',
        'Vill eleven att du rättar: be om elevens eget svar och hur hen tänkte. Säg om metoden håller',
        'och var det är värt att titta en gång till — men säg aldrig vad det rätta svaret är.',
        '',
        'HJÄLPSTEG ' + (steg + 1) + ' av 4 — ' + STEG[steg].namn + '.',
        STEG[steg].order,
        steg === 3 ? 'Du är på sista steget. Mer hjälp än så här finns inte, och det säger du rakt ut.' : '',
        tjatCount >= 2 ? 'Eleven har nu bett om svaret ' + tjatCount + ' gånger. Håll linjen vänligt och variera hur du säger nej — upprepa inte samma mening.' : '',
        '',
        bok ? 'Elevens arbetsbok "' + bok.name + '" följer med frågan. Utgå från den, hänvisa till kapitel och sidor, och säg till om något inte står i boken.'
            : 'Eleven har ingen arbetsbok uppladdad. Hjälp ändå, men säg gärna att du kan mer om boken läggs in.',
        '',
        'CANVAS — interaktiva exempel:',
        'Får du chansen att visa något som eleven kan dra i, avsluta svaret med EN rad:',
        '[[canvas:{...}]]',
        'Typerna och deras värden:',
        Canvas.beskrivning(),
        'Canvasen ska vara ett EXEMPEL med andra siffror än elevens uppgift — aldrig elevens eget tal färdiglöst.',
        'Hoppa över canvasraden när den inte tillför något. Högst en per svar.'
      ].filter(function (r) { return r !== ''; }).join('\n');
    },

    /* ---------- hjälpstegen ---------- */
    steg: function () { return App.Store.get('monni.steg', 0); },
    setSteg: function (n) { App.Store.set('monni.steg', Math.max(0, Math.min(3, n))); },
    nyUppgift: function () {
      App.Store.set('monni.steg', 0);
      App.Store.set('monni.tjat', 0);
      App.Store.set('monni.historik', []);
    },
    tjat: function () { return App.Store.get('monni.tjat', 0); },

    /* Känner igen att eleven ber om själva svaret, så att stegen inte flyttas
       fram av ett tjat och så att Monni vet om att det upprepas. */
    berOmSvar: function (text) {
      var t = ' ' + String(text || '').toLowerCase().replace(/[^\wåäöéü ]+/g, ' ') + ' ';
      var moster = [
        /\b(vad|vilket) (är|blir) svaret\b/, /\bsäg (mig )?svaret\b/, /\bbara svaret\b/,
        /\bge mig svaret\b/, /\bvad är rätt svar\b/, /\bskriv svaret\b/, /\bvisa svaret\b/,
        /\bfacit\b/,
        /\blös(a)?\s+(den|det|dem|uppgiften|talet|frågan)?\s*(åt|för)\s+mig\b/,
        /\bgör(a)?\s+(den|det|uppgiften|talet)?\s*(åt|för)\s+mig\b/,
        /\bsvara (åt|för) mig\b/, /\bräkna ut (det|den|talet) åt mig\b/, /\bvad blir det\b/
      ];
      for (var i = 0; i < moster.length; i++) if (moster[i].test(t)) return true;
      return false;
    },

    /* ---------- svarsvakten ----------
       Två kontroller på det Monni skrev, innan eleven ser det.

       1. Levererande fraser: "svaret är", "facit är", "lösningen blir" …
       2. Ekot: räknar ut elevens eget tal och letar efter resultatet i svaret,
          men bara där det står som ett levererat svar (efter = eller "blir").
          Ett resonemang som nämner samma siffra i förbifarten passerar. */
    FRASER: /(?:^|[.!?]\s|\n)[^.!?\n]*\b(svaret (?:är|blir)|rätta svaret|rätt svar är|facit (?:är|blir)|lösningen (?:är|blir)|resultatet (?:är|blir)|du (?:får|ska få) svaret)\b[^.!?\n]*[.!?]?/gi,

    /* Plockar ut ett räknetal ur elevens text och räknar ut det. null om
       det inte finns något entydigt tal att räkna på. */
    elevensTal: function (text) {
      var m = /(-?\d+(?:[.,]\d+)?)\s*([+\-*x×/:÷])\s*(-?\d+(?:[.,]\d+)?)/.exec(String(text || ''));
      if (!m) return null;
      var a = parseFloat(m[1].replace(',', '.'));
      var b = parseFloat(m[3].replace(',', '.'));
      var op = m[2];
      var v = op === '+' ? a + b
        : op === '-' ? a - b
        : (op === '*' || op === 'x' || op === '×') ? a * b
        : b === 0 ? null : a / b;
      if (v === null || !isFinite(v)) return null;
      return Math.round(v * 1000) / 1000;
    },

    KNUFF: 'Det svaret får du komma fram till själv — men jag stannar kvar och hjälper dig dit.',

    /* Modellen skriver LaTeX och markdown även när den blivit tillsagd att
       låta bli. I en chattbubbla renderas ingetdera, så "$8 \\times 7$" blir
       stående som just det. Städas bort innan eleven ser det. */
    stada: function (text) {
      return String(text || '')
        .replace(/\$\$?([^$]*)\$\$?/g, '$1')
        .replace(/\\times/g, '×').replace(/\\cdot/g, '·').replace(/\\div/g, '÷')
        .replace(/\\frac\{([^{}]*)\}\{([^{}]*)\}/g, '$1/$2')
        .replace(/\\left|\\right|\\quad|\\,/g, '')
        .replace(/\*\*([^*]+)\*\*/g, '$1')
        .replace(/(^|\s)\*([^*\n]+)\*(?=\s|$)/g, '$1$2')
        .replace(/^#{1,6}\s+/gm, '')
        .replace(/[ \t]{2,}/g, ' ')
        .trim();
    },

    vakt: function (svar, elevText) {
      var text = String(svar || '');
      var andrad = false;

      var utanFras = text.replace(this.FRASER, function (mening) {
        andrad = true;
        return (/^\n/.test(mening) ? '\n' : ' ') + Monni.KNUFF;
      });
      text = utanFras;

      var facit = this.elevensTal(elevText);
      if (facit !== null) {
        var talet = String(facit).replace('.', '[.,]');
        /* Utlösaren får stå ett par småord från talet ("blir det 56"), och en
           punkt efter talet ska inte rädda det ("Summan är 100."). */
        var levererat = new RegExp(
          '(=|\\bblir\\b|\\bär\\b|\\bsumman\\b|\\bprodukten\\b|\\bkvoten\\b|\\bdifferensen\\b)' +
          '(?:\\s+[a-zåäö]{1,7}){0,2}\\s*(?:ca\\.?\\s*)?' + talet + '(?!\\d)(?![.,]\\d)', 'i');
        if (levererat.test(text)) {
          text = text.replace(new RegExp('[^.!?\\n]*' + levererat.source + '[^.!?\\n]*[.!?]?', 'gi'), ' ' + this.KNUFF);
          andrad = true;
        }
      }
      text = text.replace(/[ \t]{2,}/g, ' ').replace(/\n{3,}/g, '\n\n').trim();
      return { text: text, andrad: andrad };
    },

    /* ---------- samtalet ---------- */
    historik: function () { return App.Store.get('monni.historik', []); },
    sparaTur: function (roll, text) {
      var h = this.historik();
      h.push({ role: roll === 'monni' ? 'model' : 'user', parts: [{ text: text }] });
      App.Store.set('monni.historik', h.slice(-12));
    },
    loggen: function () { return App.Store.get('monni.logg', []); },
    sparaLogg: function (rad) {
      var l = this.loggen();
      l.push(rad);
      App.Store.set('monni.logg', l.slice(-60));
    },

    /* Skickar elevens fråga och lämnar tillbaka ett granskat svar.
       cb(fel, {text, canvas, steg, vaktSlog}) */
    fraga: function (elevText, cb) {
      var self = this;
      var tjatade = this.berOmSvar(elevText);
      if (tjatade) App.Store.set('monni.tjat', this.tjat() + 1);
      /* Stegen flyttas fram av riktiga frågor, inte av tjat om facit. */
      var steg = this.steg();
      if (!tjatade && this.historik().length) steg = Math.min(3, steg + 1);
      this.setSteg(steg);

      App.Gemini.generate({
        prompt: elevText,
        system: this.system(steg, this.tjat()),
        history: this.historik(),
        temperature: 0.7,
        label: 'Monni: hjälp'
      }, function (err, text) {
        if (err) { cb(err); return; }
        var delat = Canvas.plocka(text);
        var granskat = self.vakt(self.stada(delat.text), elevText);
        self.sparaTur('elev', elevText);
        self.sparaTur('monni', granskat.text);
        self.sparaLogg({ t: Date.now(), fraga: elevText, steg: steg, vakt: granskat.andrad });
        App.sparaKnuff(elevText, granskat.text, steg);
        cb(null, {
          text: granskat.text,
          canvas: delat.canvas,
          steg: steg,
          vaktSlog: granskat.andrad
        });
      });
    }
  };

  global.Monni = Monni;


  /* ================== vyn ================== */
  App.registrera('monni', function (wrap, arg) {
    var bok = App.Gemini.docs.aktiv();
    App.saknasNyckel(wrap);

    var topp = App.el('div', 'row');
    var rubrik = App.el('div', 'grow');
    rubrik.appendChild(App.el('h2', 'rubrik', '💬 Monni'));
    rubrik.appendChild(App.el('p', 'ingress',
      bok ? 'Utgår från ' + bok.name : 'Ingen arbetsbok uppladdad än'));
    topp.appendChild(rubrik);
    topp.appendChild(App.button('Ny uppgift', 'ghost liten', function () {
      App.confirm('Börja om?', 'Samtalet rensas och hjälpen börjar från steg 1.', function () {
        Monni.nyUppgift();
        App.Store.set('monni.chatt', []);
        App.open('monni');
      });
    }));
    wrap.appendChild(topp);

    var regel = App.el('div', 'card');
    regel.innerHTML = '<h3>🔒 Monni säger aldrig svaret</h3>' +
      '<p class="muted">Inte om du ber snällt, inte om du ber tio gånger. Hjälpen kommer i fyra steg ' +
      'och sista steget är alltid ditt. Monni jobbar bara med text.</p>';
    wrap.appendChild(regel);

    var chat = App.el('div', null);
    chat.id = 'monni-chat';
    wrap.appendChild(chat);

    function tillBotten() {
      var v = document.getElementById('view');
      requestAnimationFrame(function () { v.scrollTop = v.scrollHeight; });
    }

    var visade = App.Store.get('monni.chatt', []);
    function bubbla(roll, text, canvas, steg) {
      var b = App.el('div', 'bubbla ' + (roll === 'elev' ? 'jag' : roll === 'system' ? 'system' : 'monni'));
      if (roll === 'monni') {
        var n = App.el('div', 'monni-namn', 'Monni');
        b.appendChild(n);
        b.appendChild(document.createTextNode(text));
        if (steg != null) {
          var m = App.el('div', 'steg-marke', 'Hjälpsteg ' + (steg + 1) + ' av 4 · ' + Monni.STEG[steg].namn);
          b.appendChild(m);
        }
      } else {
        b.textContent = text;
      }
      chat.appendChild(b);
      if (canvas) Canvas.rita(canvas, chat);
      /* Rullas det innan canvasen ritats hamnar rutan halvvägs under
         skrivraden — vänta ett varv så att höjden är känd. */
      tillBotten();
      return b;
    }

    if (!visade.length) {
      bubbla('monni', bok
        ? 'Hej! Vad har du fastnat på? Skriv uppgiften, eller vilken sida och nummer den står på.'
        : 'Hej! Lägg upp din arbetsbok under 📖 Boken så kan jag hjälpa dig med just dina uppgifter. Du kan fråga mig ändå.',
        null, null);
    } else {
      visade.forEach(function (r) { bubbla(r.roll, r.text, r.canvas, r.steg); });
    }

    var fot = App.el('div', null);
    fot.id = 'monni-fot';
    var snabbval = App.el('div', 'snabbrad');
    [['Jag fastnade här', 'Jag har fastnat på den här uppgiften: '],
     ['Förklara enklare', 'Kan du förklara det där enklare?'],
     ['Visa ett exempel', 'Kan du visa ett liknande exempel med andra siffror?'],
     ['Ge mig en canvas', 'Kan du visa ett interaktivt exempel jag kan dra i?']].forEach(function (v) {
      snabbval.appendChild(App.button(v[0], 'ghost liten', function () {
        inp.value = v[1];
        inp.focus();
      }));
    });
    fot.appendChild(snabbval);

    var inrad = App.el('div', 'inrad');
    var inp = App.el('textarea', 'falt');
    inp.id = 'monni-in';
    inp.placeholder = arg && arg.fraga ? arg.fraga : 'Vad har du fastnat på?';
    if (arg && arg.text) inp.value = arg.text;
    inrad.appendChild(inp);
    var skicka = App.button('➤', '', function () { skickaFraga(); });
    skicka.setAttribute('aria-label', 'Skicka');
    inrad.appendChild(skicka);
    fot.appendChild(inrad);
    wrap.appendChild(fot);

    function skickaFraga() {
      var text = inp.value.trim();
      if (!text) return;
      inp.value = '';
      skicka.disabled = true;
      bubbla('elev', text);
      var loggat = App.Store.get('monni.chatt', []);
      loggat.push({ roll: 'elev', text: text });
      App.Store.set('monni.chatt', loggat.slice(-40));

      var vantar = App.el('div', 'bubbla monni skriver');
      vantar.innerHTML = '<span></span><span></span><span></span>';
      chat.appendChild(vantar);
      tillBotten();

      Monni.fraga(text, function (err, svar) {
        chat.removeChild(vantar);
        skicka.disabled = false;
        if (err) { bubbla('system', err); return; }
        bubbla('monni', svar.text, svar.canvas, svar.steg);
        if (svar.vaktSlog) bubbla('system', '🔒 Svarsvakten tog bort ett färdigt svar ur Monnis text.');
        var l = App.Store.get('monni.chatt', []);
        l.push({ roll: 'monni', text: svar.text, canvas: svar.canvas, steg: svar.steg });
        App.Store.set('monni.chatt', l.slice(-40));
      });
    }
    inp.addEventListener('keydown', function (e) {
      if (e.key === 'Enter' && !e.shiftKey) { e.preventDefault(); skickaFraga(); }
    });
  });
})(window);

File: ios/SandoElev/Modell/Monni.swift
import Foundation

/// Monni — reglerna, inte gränssnittet.
///
/// Samma tre lager som i webbappen (se ../../../MONNI.md), för en systemprompt
/// ensam är en önskan och inte en spärr:
///
///   1. systemprompt(steg:tjat:) — vad Monni gör i stället för att svara
///   2. Steg — hjälpen växer i fyra steg och stannar vid det fjärde
///   3. vakt(_:elevText:) — en lokal kontroll av det Monni faktiskt skrev
///
/// Vakten är testad i SandoElevTests med samma tabeller som webbappens
/// självtest: sju svar som ska stoppas och sex förklaringar som absolut inte
/// får stoppas. En vakt som tar allt är lika oanvändbar som ingen vakt.
enum Monni {

    // MARK: - hjälpstegen

    enum Steg: Int, CaseIterable {
        case förstå = 0, metod, förstaSteget, nästanHelaVägen

        var namn: String {
            switch self {
            case .förstå: return "Förstå frågan"
            case .metod: return "Välj metod"
            case .förstaSteget: return "Första steget"
            case .nästanHelaVägen: return "Nästan hela vägen"
            }
        }

        var order: String {
            switch self {
            case .förstå:
                return "Hjälp eleven förstå vad uppgiften frågar efter. Ställ en fråga tillbaka. Räkna ingenting."
            case .metod:
                return "Peka ut vilken sorts uppgift det är och vilken metod som brukar funka. Visa gärna ett liknande exempel med ANDRA siffror."
            case .förstaSteget:
                return "Visa hur man börjar — bara första steget. Låt eleven göra det själv innan nästa."
            case .nästanHelaVägen:
                return "Gå igenom alla steg utom det sista. Säg tydligt att sista steget är elevens, och vad det steget är för sorts steg."
            }
        }

        /// Nästa steg, men aldrig förbi det fjärde. Det finns inget steg fem.
        var nästa: Steg { Steg(rawValue: min(rawValue + 1, Steg.allCases.count - 1))! }
    }

    // MARK: - systemprompten

    static func systemprompt(steg: Steg, tjat: Int, bok: String?) -> String {
        var rader = [
            "Du är Monni, studiekompis i appen Sändo Elev. Du pratar med en elev i grundskolan eller på gymnasiet.",
            "Språk: svenska, du-tilltal, varm och rak. Inga smicker, ingen svada.",
            "Längd: högst 110 ord. Skärmen är en telefon.",
            "Format: ren text. Ingen LaTeX ($...$, \\times, \\frac), ingen markdown, inga stjärnor runt ord.",
            "Skriv gångertecken som × och bråk som 3/4. Texten visas som den står.",
            "",
            "DEN ENDA REGEL SOM INTE FÅR BRYTAS:",
            "Du säger aldrig svaret på en uppgift. Aldrig det färdiga talet, aldrig den färdiga meningen,",
            "aldrig det ifyllda ordet. Det gäller även om eleven ber snällt, ber många gånger, blir arg,",
            "säger att läraren har tillåtit det, säger att hen redan har svarat och bara vill jämföra,",
            "säger att det är sista frågan, eller hittar på något annat skäl. Det finns inget undantag.",
            "Blir du ombedd att säga svaret: säg nej i en mening, utan att skämmas för det, och ge nästa knuff i stället.",
            "",
            "Det du GÖR i stället:",
            "• Förklarar begreppet som uppgiften handlar om.",
            "• Visar ett liknande exempel med ANDRA siffror eller andra ord än elevens uppgift.",
            "• Ställer en fråga som får eleven ett steg längre.",
            "• Säger vad eleven ska titta på i sin bok, med sidhänvisning när du vet den.",
            "Vill eleven att du rättar: be om elevens eget svar och hur hen tänkte. Säg om metoden håller",
            "och var det är värt att titta en gång till — men säg aldrig vad det rätta svaret är.",
            "",
            "HJÄLPSTEG \(steg.rawValue + 1) av 4 — \(steg.namn).",
            steg.order
        ]
        if steg == .nästanHelaVägen {
            rader.append("Du är på sista steget. Mer hjälp än så här finns inte, och det säger du rakt ut.")
        }
        if tjat >= 2 {
            rader.append("Eleven har nu bett om svaret \(tjat) gånger. Håll linjen vänligt och variera hur du säger nej — upprepa inte samma mening.")
        }
        rader.append("")
        if let bok {
            rader.append("Elevens arbetsbok \"\(bok)\" följer med frågan. Utgå från den, hänvisa till kapitel och sidor, och säg till om något inte står i boken.")
        } else {
            rader.append("Eleven har ingen arbetsbok uppladdad. Hjälp ändå, men säg gärna att du kan mer om boken läggs in.")
        }
        return rader.joined(separator: "\n")
    }

    // MARK: - tjatdetektorn

    private static let tjatmönster = [
        #"\b(vad|vilket) (är|blir) svaret\b"#,
        #"\bsäg (mig )?svaret\b"#,
        #"\bbara svaret\b"#,
        #"\bge mig svaret\b"#,
        #"\bvad är rätt svar\b"#,
        #"\bskriv svaret\b"#,
        #"\bvisa svaret\b"#,
        #"\bfacit\b"#,
        #"\blös(a)? (den |det |dem |uppgiften |talet |frågan )?(åt|för) mig\b"#,
        #"\bgör(a)? (den |det |uppgiften |talet )?(åt|för) mig\b"#,
        #"\bsvara (åt|för) mig\b"#,
        #"\bräkna ut (det|den|talet) åt mig\b"#,
        #"\bvad blir det\b"#
    ]

    /// Känner igen att eleven ber om själva svaret, så att hjälpstegen inte
    /// flyttas fram av ett tjat — annars hade fyra "säg svaret" räckt för att
    /// pressa fram den största hjälpen appen har.
    static func berOmSvar(_ text: String) -> Bool {
        let t = " " + text.lowercased()
            .replacingOccurrences(of: #"[^\wåäöéü ]+"#, with: " ", options: .regularExpression) + " "
        return tjatmönster.contains { t.range(of: $0, options: .regularExpression) != nil }
    }

    // MARK: - svarsvakten

    static let knuff = "Det svaret får du komma fram till själv — men jag stannar kvar och hjälper dig dit."

    private static let levererandeFraser =
        #"(?:^|[.!?]\s|\n)[^.!?\n]*\b(svaret (?:är|blir)|rätta svaret|rätt svar är|facit (?:är|blir)|lösningen (?:är|blir)|resultatet (?:är|blir)|du (?:får|ska få) svaret)\b[^.!?\n]*[.!?]?"#

    /// Räknar ut elevens eget tal, om frågan innehåller ett entydigt sådant.
    static func elevensTal(_ text: String) -> Double? {
        let mönster = #"(-?\d+(?:[.,]\d+)?)\s*([+\-*x×/:÷])\s*(-?\d+(?:[.,]\d+)?)"#
        guard let re = try? NSRegularExpression(pattern: mönster),
              let m = re.firstMatch(in: text, range: NSRange(text.startIndex..., in: text)),
              let r1 = Range(m.range(at: 1), in: text),
              let r2 = Range(m.range(at: 2), in: text),
              let r3 = Range(m.range(at: 3), in: text),
              let a = Double(text[r1].replacingOccurrences(of: ",", with: ".")),
              let b = Double(text[r3].replacingOccurrences(of: ",", with: "."))
        else { return nil }

        let v: Double?
        switch text[r2] {
        case "+": v = a + b
        case "-": v = a - b
        case "*", "x", "×": v = a * b
        default: v = b == 0 ? nil : a / b
        }
        guard let v, v.isFinite else { return nil }
        return (v * 1000).rounded() / 1000
    }

    /// Läser det Monni skrev innan eleven ser det. Två kontroller:
    /// levererande fraser, och ekot av elevens eget tal där det står som ett
    /// levererat svar. Returnerar texten och om något byttes ut.
    static func vakt(_ svar: String, elevText: String) -> (text: String, ändrad: Bool) {
        var text = svar
        var ändrad = false

        if let re = try? NSRegularExpression(pattern: levererandeFraser, options: [.caseInsensitive]) {
            let n = NSRange(text.startIndex..., in: text)
            if re.firstMatch(in: text, range: n) != nil {
                text = re.stringByReplacingMatches(in: text, range: n, withTemplate: " " + knuff)
                ändrad = true
            }
        }

        if let facit = elevensTal(elevText) {
            let talet = tallitteral(facit)
            // Utlösaren får stå ett par småord från talet ("blir det 56"), och
            // en punkt efter talet ska inte rädda det ("Summan är 100.").
            let utlösare = #"(=|\bblir\b|\bär\b|\bsumman\b|\bprodukten\b|\bkvoten\b|\bdifferensen\b)"#
                + #"(?:\s+[a-zåäö]{1,7}){0,2}\s*(?:ca\.?\s*)?"# + talet + #"(?!\d)(?![.,]\d)"#
            let hela = #"[^.!?\n]*"# + utlösare + #"[^.!?\n]*[.!?]?"#
            if let re = try? NSRegularExpression(pattern: hela, options: [.caseInsensitive]) {
                let n = NSRange(text.startIndex..., in: text)
                if re.firstMatch(in: text, range: n) != nil {
                    text = re.stringByReplacingMatches(in: text, range: n, withTemplate: " " + knuff)
                    ändrad = true
                }
            }
        }

        text = text
            .replacingOccurrences(of: #"[ \t]{2,}"#, with: " ", options: .regularExpression)
            .replacingOccurrences(of: #"\n{3,}"#, with: "\n\n", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
        return (text, ändrad)
    }

    /// "56" respektive "3[.,]5" — kommatecken och punkt ska båda träffa.
    private static func tallitteral(_ v: Double) -> String {
        let s = v == v.rounded() ? String(Int(v)) : String(v)
        return s.replacingOccurrences(of: ".", with: "[.,]")
    }

    // MARK: - städning

    /// Modellen skriver LaTeX och markdown även när den blivit tillsagd att låta
    /// bli. En chattbubbla renderar ingetdera, så "$8 \times 7$" blir stående
    /// som just det.
    static func städa(_ text: String) -> String {
        var t = text
        let byten: [(String, String)] = [
            (#"\$\$?([^$]*)\$\$?"#, "$1"),
            (#"\\times"#, "×"), (#"\\cdot"#, "·"), (#"\\div"#, "÷"),
            (#"\\frac\{([^{}]*)\}\{([^{}]*)\}"#, "$1/$2"),
            (#"\\left|\\right|\\quad|\\,"#, ""),
            (#"\*\*([^*]+)\*\*"#, "$1"),
            (#"(?m)^#{1,6}\s+"#, ""),
            (#"[ \t]{2,}"#, " ")
        ]
        for (mönster, ersättning) in byten {
            t = t.replacingOccurrences(of: mönster, with: ersättning, options: .regularExpression)
        }
        return t.trimmingCharacters(in: .whitespacesAndNewlines)
    }
}
